docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx

```diff
@@ -0,0 +1,1478 @@
+
+These are the mappings from the built in scalar TypeSpec types.
+We add `encode` onto these types, telling us how to encode when sending to the service.
+
+There is a one-to-one mapping between the TypeSpec scalar kinds and the `SdkBuiltInKinds`.
+
+#### Interface
+
+```ts
+export interface SdkBuiltInType extends SdkTypeBase {
+  kind: SdkBuiltInKinds;
+  encode: string;
+}
+```
+
+### SdkDateTimeType
+
+```ts
+interface SdkDatetimeTypeBase extends SdkTypeBase {
+  encode: DateTimeKnownEncoding;
+  // what we send over the wire. Often it's string
+  wireType: SdkBuiltInType;
+}
+
+interface SdkUtcDatetimeType extends SdkDatetimeTypeBase {
+  kind: "utcDateTime";
+}
+
+interface SdkOffsetDatetimeType extends SdkDatetimeTypeBase {
+  kind: "offsetDateTime";
+}
+```
+
+### SdkDurationType
+
+```ts
+interface SdkDurationType extends SdkTypeBase {
+  kind: "duration";
+  encode: DurationKnownEncoding;
+  // What we send over the wire. It's usually either a string or a float
+  wireType: SdkBuiltInType;
+}
+```
+
+### SdkArrayType
+
+```ts
+interface SdkArrayType extends SdkTypeBase {
+  kind: "array";
+  valueType: SdkType;
+}
+```
+
+### SdkTupleType
+
+```ts
+interface SdkTupleType extends SdkTypeBase {
+  kind: "tuple";
+  values: SdkType[];
+}
+```
+
+### SdkDictionaryType
+
+```ts
+interface SdkDictionaryType extends SdkTypeBase {
+  kind: "dict";
+  keyType: SdkType; // currently can only be string
+  valueType: SdkType;
+}
+```
+
+### SdkEnumType
+
+```ts
+export interface SdkEnumType extends SdkTypeBase {
+  kind: "enum";
+  name: string;
+  // If it's an anonymous enum, this is the name we generate
+  generatedName?: string;
+  valueType: SdkBuiltInType;
+  values: SdkEnumValueType[];
+  isFixed: boolean;
+  description?: string;
+  isFlags: boolean;
+  usage: UsageFlags;
+  access: AccessFlags;
+  crossLanguageDefinitionId: string;
+  apiVersions: string[];
+}
+```
+
+### SdkEnumValueType
+
+```ts
+export interface SdkEnumValueType extends SdkTypeBase {
+  kind: "enumvalue";
+  name: string;
+  value: string | number;
+  enumType: SdkEnumType;
+  valueType: SdkType;
+  description?: string;
+  details?: string;
+}
+```
+
+### SdkConstantType
+
+```ts
+export interface SdkConstantType extends SdkTypeBase {
+  kind: "constant";
+  value: string | number | boolean | null;
+  valueType: SdkBuiltInType;
+}
+```
+
+### SdkUnionType
+
+```ts
+export interface SdkUnionType extends SdkTypeBase {
+  name?: string;
+  generatedName?: string;
+  kind: "union";
+  values: SdkType[];
+}
+```
+
+### SdkModelType
+
+```ts
+export interface SdkModelType extends SdkTypeBase {
+  kind: "model";
+  // purposely can also be header / query params for fidelity with TypeSpec
+  properties: SdkModelPropertyType[];
+  name: string;
+  isFormDataType: boolean;
+  isError: boolean;
+  generatedName?: string;
+  description?: string;
+  details?: string;
+  access?: AccessFlags;
+  usage: UsageFlags;
+  additionalProperties?: SdkType;
+  discriminatorValue?: string;
```

Comment:
should we make a separate `discriminatedModel` type with the discriminator goop on it?